### PR TITLE
Add LUTs generation on fly if ktx2 is not enabled

### PR DESCRIPTION
**Pull request by bugsweeper**
Add LUTs generation on fly if ktx2 is not enabled

# Objective

- Partially addresses #13728  .
- If default features are disabled some subcrates has implicit but not explicit dependencies.
For example `bevy_core_pipeline` adds `SmaaPlugin` which wants `ktx2` textures, `ktx2` uses `zstd` for supercompression

## Solution

- Added feature `smaa_luts`. If enabled SMAA uses LUTs from ktx2 files, if disabled SMAA uses LUTs generated on fly. Generators are based on on Jorge Jimenez code for [AreaTex](https://github.com/iryoku/smaa/blob/master/Scripts/AreaTex.py) and [SearchTex](https://github.com/iryoku/smaa/blob/master/Scripts/SearchTex.py)

## Testing

- I ran reproducing example from #13728 with enabled bevy_pbr and got no panics
- I made comparison app (which compares data from ktx2 luts and from generators) that runs with no panics
```
fn main() {
    let mut app = App::new();
    app.add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .add_systems(Update, check)
        .run();
}

#[derive(Resource, Default)]
struct Luts {
    search: Handle<Image>,
    area: Handle<Image>,
}

fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.insert_resource(Luts {
        search: asset_server.load("/some/path/to/bevy_core_pipeline/src/smaa/SMAASearchLUT.ktx2"),
        area: asset_server.load("/some/path/to/bevy_core_pipeline/src/smaa/SMAAAreaLUT.ktx2"),
    })
}

fn check(
    mut processed: Local<u8>,
    luts: Res<Luts>,
    images: Res<Assets<Image>>,
    mut events: EventReader<AssetEvent<Image>>,
    mut app_exit_writer: EventWriter<AppExit>,
) {
    for event in events.read() {
        if event.is_loaded_with_dependencies(luts.search.id()) {
            if let Some(search) = images.get(luts.search.id()) {
                let generated = search_data()
                    .into_iter()
                    .map(|color| {
                        (Srgba::gamma_function(color as f32 / 255.) * 255.).floor()
                            as u8
                    })
                    .collect::<Vec<_>>();
                // Search look up tables are exactly the same
                assert_eq!(search.data, generated);
                *processed += 1;

                continue;
            }
        }
        if event.is_loaded_with_dependencies(luts.area.id()) {
            if let Some(area) = images.get(luts.area.id()) {
                let generated = area_data()
                    .into_iter()
                    .map(|color| {
                        (Srgba::gamma_function(color as f32 / 255.) * 255.).floor()
                            as u8
                    })
                    .collect::<Vec<_>>();
                assert_eq!(
                    area.data
                        .iter()
                        .zip(generated.iter())
                        .fold(0, |sum, (&goal, &generated)| {
                            let diff = goal.abs_diff(generated);
                            // Some pixels differs in color by 1-2 color points
                            assert!(diff < 3);
                            sum + diff as usize
                        }),
                    // Sum of diffs is only 582 from 179200 values
                    // diffs are mostly on borders of subsamples
                    // because compression algorithms usualy don't like high contrast
                    582
                );

                for (goal, &generated) in area.data.iter().zip(generated.iter()) {
                    // Some pixels differs in color by 1-2 color points
                    assert!(goal.abs_diff(generated) < 3);
                }

                *processed += 1;
            }
        }
    }

    if *processed == 2 {
        app_exit_writer.send(AppExit::Success);
    }
}
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -104,6 +104,8 @@ bevy_core_pipeline = [
   "bevy_internal/bevy_core_pipeline",
   "bevy_asset",
   "bevy_render",
+  "bevy_pbr",
```
Comment (alice-i-cecile) [inline Cargo.toml:107]: I don't like pulling in `bevy_pbr` here. What's causing the failure?

Comment (alice-i-cecile) [reply Cargo.toml:107]: > which wants bevy_pbr for Opaque3dPrepass bindings initialization

Comment (alice-i-cecile) [reply Cargo.toml:107]: Let me pull in rendering folks, but I think the correct solution is to actually move that code.

Comment (bugsweeper) [reply Cargo.toml:107]: I said about [this initialization](https://github.com/bevyengine/bevy/blob/ecdd1624f302c5f71aaed95b0984cbbecf8880b7/crates/bevy_pbr/src/prepass/mod.rs#L151) in description

Comment (bugsweeper) [reply Cargo.toml:107]: > I think the correct solution is to actually move that code.

Which code are you proposing to move? Core3dPlugin from bevy_core_pipeline to bevy_pbr?

**Comment by mockersf** [comment]
`ktx2` should not enable `zstd`, and `bevy_core_pipeline` should not enable `bevy_pbr` or `ktx2`

**Comment by bugsweeper** [comment]
> `ktx2` should not enable `zstd`, and `bevy_core_pipeline` should not enable `bevy_pbr` or `ktx2`

As I mentioned in PR there are dependencies, should SmaaPlugin, Core3dPlugin become optional, or everyone must enable all necessary features manualy?

**Comment by alice-i-cecile** [comment]
> > `ktx2` should not enable `zstd`, and `bevy_core_pipeline` should not enable `bevy_pbr` or `ktx2`
> 
> As I mentioned in PR there are dependencies, should SmaaPlugin, Core3dPlugin become optional, or everyone must enable all necessary features manualy?

We should move the offending code out of `bevy_pbr` and into `bevy_core_pipeline`.

**Comment by mockersf** [comment]
for ktx2 / ztsd features, like tonemapping LUTs being behind a feature `tonemapping_luts`, SMAA LUTs should be behind a new feature `smaa_luts`

**Comment by bugsweeper** [comment]
Generator for search LUT generates exactly the same table, generator of area LUT generates table which absolutely differs up to 2. I think it's due to compression loss in ktx2 (Generators are even more precise)

**Comment by bugsweeper** [comment]
At this moment only bevy_core_pipeline dependences from ktx2/zstd  are fixed

**Comment by bugsweeper** [comment]
> we should move the offending code out of bevy_pbr and into bevy_core_pipeline

@alice-i-cecile Sorry, I tried to move out code from bevy_pbr to bevy_core_pipeline, that code appeared to be very dependent.  As I mentioned earlier `bevy_pbr` adds plugins:
```
BinnedRenderPhasePlugin::<Opaque3dPrepass, MeshPipeline>::default(),
BinnedRenderPhasePlugin::<AlphaMask3dPrepass, MeshPipeline>::default(),
```
The journey started from `MeshPipeline`, and one by one in a chain it forces me to move `MeshPipelineViewLayouts`, `MeshLayouts`, `MeshPipelineViewLayout`, `MeshUniform` and so on, as far I was going as more I was feeling that I'm doing something wrong.
I propose to make this change focused on SMAA LUTs and drop connection to issue #13728 , because fix is not complete and we should not close issue automaticaly

**Comment by alice-i-cecile** [comment]
Sounds good, let's land something more scoped.

**Comment by bugsweeper** [comment]
> added [S-Waiting-on-Author](https://github.com/bevyengine/bevy/labels/S-Waiting-on-Author) and removed [S-Needs-Review](https://github.com/bevyengine/bevy/labels/S-Needs-Review)

Changes to replace loading SMAA LUT from ktx2 files are done. This PR is no longer waiting for the author.

**Comment by bugsweeper** [comment]
Should I make unit/integration test (maybe ignored by default) from test code in PR description?

**Comment by alice-i-cecile** [comment]
I'm on the fence about an integration test. On the one hand, this is hard to avoid breaking. On the other hand, I worry about the test being brittle. I'd prefer to set up integration tests for feature flags in a separate PR, since they're quite complex to design.

**Comment by JMS55** [comment]
My 2cents: Generating LUTs is a lot of code for no good reason. SMAA just shouldn't be usable without ktx2.

Furthermore, we don't need to put SMAA behind a new smaa cfg. We can leave the majority of the code exposed without a feature flag, and simply put the LUT load behind a cfg for ktx2, and panic if ktx2 is disabled.

**Pull request closed without merging**